Duplicati/Library/Main/Operation/DeleteHandler.cs

```diff
@@ -173,45 +174,92 @@ where toDelete.Contains(n.Item2)
         /// <param name="allBackups">The list of backups that can be deleted</param>
-        private DateTime[] GetFilesetsToDelete(DateTime[] allBackups)
+        private DateTime[] GetFilesetsToDelete(Database.LocalDeleteDatabase db, DateTime[] allBackups)
         {
             if (allBackups.Length == 0)
+            {
                 return allBackups;
+            }
+
+            Logging.Log.WriteDryrunMessage(LOGTAG, "***** FileSets", $"List of backup timestamps: {string.Join(", ", allBackups.Select(x => x.ToString()))}");
+
 
             if (allBackups.Select(x => x.ToUniversalTime()).Distinct().Count() != allBackups.Length)
-                throw new Exception(string.Format("List of backup timestamps contains duplicates: {0}", string.Join(", ", allBackups.Select(x => x.ToString()))));
+            {
+                throw new Exception($"List of backup timestamps contains duplicates: {string.Join(", ", allBackups.Select(x => x.ToString()))}");
+            }
 
             List<DateTime> toDelete = new List<DateTime>();
 
-            // Remove backups explicitely specified via option
+            // Remove backups explicitly specified via option
             var versions = m_options.Version;
             if (versions != null && versions.Length > 0)
+            {
                 foreach (var ix in versions.Distinct())
+                {
                     if (ix >= 0 && ix < allBackups.Length)
+                    {
                         toDelete.Add(allBackups[ix]);
+                    }
+                }
+            }
 
             // Remove backups that are older than date specified via option
             var keepTime = m_options.KeepTime;
             if (keepTime.Ticks > 0)
+            {
                 toDelete.AddRange(allBackups.SkipWhile(x => x >= keepTime));
+            }
 
-            // Remove backups via rentention policy option
-            toDelete.AddRange(ApplyRetentionPolicy(allBackups));
+            // Remove backups via retention policy option
+            toDelete.AddRange(ApplyRetentionPolicy(db, allBackups));
 
-            // Check how many backups will be remaining after the previous steps
+            // Check how many full backups will be remaining after the previous steps
             // and remove oldest backups while there are still more backups than should be kept as specified via option
             var backupsRemaining = allBackups.Except(toDelete).ToList();
             var keepVersions = m_options.KeepVersions;
-            if (keepVersions > 0 && keepVersions < backupsRemaining.Count())
-                toDelete.AddRange(backupsRemaining.Skip(keepVersions));
-
-            var toDeleteDistinct = toDelete.Distinct().OrderByDescending(x => x.ToUniversalTime()).AsEnumerable();
+            var versionsKeptCount = 0;
+            if (keepVersions > 0 && keepVersions < backupsRemaining.Count)
+            {
+                // skip full versions for the count of keepVersions

```

Comment:
This comment is a little confusing to me.  Are we actually skipping partial versions when incrementing `versionsKeptCount`?

Reply:
Partials are skipped and considered to be part of the proceeding full backup. They are to be considered together a complete backup.

Reply:
That makes sense, but the confusing part to me is the comment says "skip _full_ versions"?

Reply:
keepVersion is the count of the full backups and does not include the partial backups. And backupsRemaining is also only the full backups. 

Partials don to get removed on their own, only when the matching full backup is removed.

Should I add "full" to the variables to make it more understandable?

Reply:
> Should I add "full" to the variables to make it more understandable?

No, perhaps I am just confused.  At a quick glance, I was expecting the comment to be
```
// skip partial versions for the count of keepVersions
```
I will think about this some more...

Reply:
It's not skipping partials at all. It is determining which full backups to keep.

Reply:
Ah, I get it now.  Thanks for the explanation.

Reply:
No problem. I think renaming the variables will help.